src/endpoint/locality.rs

```diff
@@ -277,34 +278,79 @@ impl LocalitySet {
     }
 
     pub fn merge(&mut self, cluster: &Self) {
+        use std::collections::hash_map::Entry;
+
         for (key, value) in &cluster.0 {
-            if tracing::enabled!(tracing::Level::INFO) {
-                let span = tracing::info_span!(
-                    "applied_locality",
-                    locality = &*key
-                        .as_ref()
-                        .map(|locality| locality.colon_separated_string())
-                        .unwrap_or_else(|| String::from("<none>"))
-                );
-
-                let _entered = span.enter();
-                if value.endpoints.is_empty() {
-                    tracing::info!("removing all endpoints");
-                } else if let Some(original_endpoints) = self.0.get(key) {
-                    for endpoint in &original_endpoints.endpoints {
-                        if !value.endpoints.contains(endpoint) {
-                            tracing::info!(%endpoint.address, ?endpoint.metadata.known.tokens, "removing endpoint");
+            let span = tracing::info_span!(
+                "applied_locality",
+                locality = &*key
+                    .as_ref()
+                    .map(|locality| locality.colon_separated_string())
+                    .unwrap_or_else(|| String::from("<none>"))
+            );
+
+            let _entered = span.enter();
+
+            match self.0.entry(key.clone()) {
+                // The eviction logic is as follows:
+                //
+                // If an endpoint already exists:
+                // - If `sessions` is zero then it is dropped.
+                // If that endpoint exists in the new set:
+                // - Its metadata is replaced with the new set.
+                // Else the endpoint remains.
+                //
+                // This will mean that updated metadata such as new tokens
+                // will be respected, but we will still retain older
+                // endpoints that are currently actively used in a session.
+                Entry::Occupied(entry) => {
```

Comment:
Is this dangerous? Basically if an endpoint is getting a malicious traffic and the solution the end user wants to implement is to drop the endpoint entirely to protect the internal resources -- doesn't this mean that that won't work while the malicious traffic is continuing?

Reply:
> Basically if an endpoint is getting a malicious traffic and the solution the end user wants to implement is to drop the endpoint entirely to protect the internal resources -- doesn't this mean that that won't work while the malicious traffic is continuing?

Well yeah, that is why we shouldn't accept this version. I'll write up my response to your other questions/proposed solution in a separate comment, but I think what we need is to be able to have different verbs here. We need to be able to distinct between "services are disrupted or a bug has caused a endpoint to be lost in the chain" where we want the load balancer to retain that endpoint because we don't want player's to be dropped mid-game, and "stop sending traffic to this endpoint because we want to stop traffic to it".